Add a migrations command

ausi opened the pull request:
See https://github.com/contao/core-bundle/issues/1223#issuecomment-527111644

Todo:
- [x] Tests
- [x] Use migrations service in the install tool

ausi committed:
Add migrations command

ausi committed:
Execute runonce files and database migration

ausi committed:
Add complete command line option

ausi committed:
Add database migration classes

ausi committed:
Coding style

ausi committed:
Initialize framework for runonce files

ausi committed:
Rename complete answer

ausi committed:
Fix database migration scripts

xchs commented on core-bundle/src/Command/MigrateCommand.php:
```diff
@@ -0,0 +1,266 @@
+
+            $this->io->listing($commandsByHash);
+
+            $options = ['yes', 'yes, with delets', 'no'];
```
```suggestion
            $options = ['yes', 'yes, with deletes', 'no'];
```

fritzmg replied:
I noticed that during they keynote too 😁

xchs commented on core-bundle/src/Command/MigrateCommand.php:
```diff
@@ -0,0 +1,266 @@
+
+            $count = 0;
+
+            foreach ($this->getCommandHashes($commands, 'yes, with delets' === $answer) as $hash) {
```
```suggestion
            foreach ($this->getCommandHashes($commands, 'yes, with deletes' === $answer) as $hash) {
```

ausi committed:
Fix typo

Co-Authored-By: Hannes <xchs@gmx-topmail.de>

ausi committed:
Fix typo

Co-Authored-By: Hannes <xchs@gmx-topmail.de>

ausi committed:
Merge branch master into feature/migrate-command

Conflicts:
	core-bundle/src/ContaoCoreBundle.php
	core-bundle/src/DependencyInjection/ContaoCoreExtension.php
	core-bundle/src/Resources/config/services.yml
	installation-bundle/src/Database/Version330Update.php
	installation-bundle/src/Database/Version350Update.php
	installation-bundle/src/Database/Version400Update.php
	installation-bundle/src/Database/Version410Update.php
	installation-bundle/src/Database/Version430Update.php
	installation-bundle/src/Database/Version440Update.php
	installation-bundle/src/Database/Version447Update.php
	installation-bundle/src/Database/Version450Update.php
	installation-bundle/src/Database/Version460Update.php
	installation-bundle/src/Database/Version470Update.php
	installation-bundle/src/Database/Version480Update.php
	installation-bundle/src/Resources/config/services.yml

ausi committed:
Fix tests

ausi committed:
Add SQL execution section headline

ausi committed:
Use migrations service in the install tool

ausi committed:
Run the CS fixer

ausi committed:
Add tests

ausi committed:
Merge branch 'master' into feature/migrate-command

ausi committed:
Improve command test

ausi committed:
Fix tests

ausi committed:
Improve tests

ausi committed:
Coding style

Toflar commented on core-bundle/src/Command/MigrateCommand.php:
```diff
@@ -0,0 +1,272 @@
+    private function getRunOnceFiles(): array
+    {
+        try {
+            $files = $this->fileLocator->locate('config/runonce.php', null, false);
```
Do you think we should add a warning/deprecation message so users can inform the developers about our new migrations framework? 🎉

ausi replied:
Agree

ausi replied:
ec79cc14865c923fa1f1dae83b31e1d2ed6ea995

Toflar commented on core-bundle/tests/Command/MigrateCommandTest.php:
```diff
@@ -0,0 +1,217 @@
+
+    public function testExecutesSchemaDiff(): void
+    {
+        $installer = $this->createMock(Installer::class);
```
Either the `contao/installation-bundle` is currently a hard dependency of the core-bundle and then the argument doesn't need to be optional in the `Contao\CoreBundle\Command\MigrateCommand` service or this test should be skipped if `!class_exists(Installer::class)`. Or did I miss anything here?

ausi replied:
The command checks if the installer was provided in https://github.com/contao/contao/pull/709/files#diff-b878e4d8368ed6f024ba9eeafc45f150R188
So the schema diff only works with the installation bundle, but the migrations can still be executed without it.

ausi replied:
I should probably improve the error message to make it clear that the installation-bundle has to be installed for the schema diff to work.

Toflar replied:
Which isn't really what we want because it should be core functionality. But I'm okay with that for now. As we want to get rid of the installation bundle one day anway. So let's go step by step here :)

ausi replied:
39bc97273ac36c9a7e4b576765b48acc45388cdf

Toflar reviewed (commented):
Only a few minor nitpicks! So much looking forward to this! 🕺

fritzmg commented:
@ausi please also provide a PR for the documentation :)

ausi committed:
Deprecate runonce files

ausi committed:
Improve missing installer message

Toflar reviewed (dismissed):
Awesome! 🎉

aschempp commented on core-bundle/src/Command/MigrateCommand.php:
```diff
@@ -0,0 +1,278 @@
+                $this->io->writeln(' * '.$result->getMessage());
+
+                if (!$result->isSuccessful()) {
+                    $this->io->error('Migration failed');
```
if a migration fails, can I tell which one it was? And do we continue with all other migrations (given they have priority)?

ausi replied:
> can I tell which one it was?

Yes, the message of the result gets written to the output three lines above.

> And do we continue with all other migrations

Yes, we do, as long as the migration fails “nicely” by returning an unseccessful result. If the migration wants to cancel the whole process it should throw an exception.

aschempp commented on core-bundle/src/Command/MigrateCommand.php:
```diff
@@ -0,0 +1,278 @@
+                $this->io->writeln(' * Executed runonce file: '.$file);
+            }
+
+            $this->io->success('Executed migrations.');
```
This does not sound like a confirmation to me. Should it be *Executed X migrations* (with the number) or something like *All migrations executed successfully.* ?

ausi replied:
Agree, I will add the count.

ausi replied:
Done in f81820a0efa747eda89ad2b7ee98faf603a58640

aschempp commented on core-bundle/src/Command/MigrateCommand.php:
```diff
@@ -0,0 +1,278 @@
+    private function executeSchemaDiff(bool $completeOption): bool
+    {
+        if (null === $this->installer) {
+            $this->io->error('Service contao.installer of contao/installation-bundle not found. The installation bundle needs to be installed in order to execute schema diff migrations.');
```
are we planning to get away from the installer? Asking because the schema diff can be done through Doctrine, but as far as I remember the Installer might add additional stuff (like collation migration?).

ausi replied:
AFAIK we are planning to get away from the installer, but this does not need to be part of this PR I think.

aschempp commented on core-bundle/src/Command/MigrateCommand.php:
```diff
@@ -0,0 +1,278 @@
+
+            $options = ['yes', 'yes, with deletes', 'no'];
+
+            if ($completeOption) {
```
this code is confusing to me, can we maybe rewrite it? I would rename the argument to something like `alwaysDelete` and only add the `yes` option if not set? or just generate the options directly in an if-else statement.

ausi replied:
Done in 84b89509580a3d5c6414989df87503b560a589b7

aschempp commented on core-bundle/src/Command/MigrateCommand.php:
```diff
@@ -0,0 +1,278 @@
+
+            foreach ($commands as $category => $commandsByHash) {
+                foreach ($commandsByHash as $hash => $command) {
+                    if ('DROP' === $category && false === strpos($command, 'DROP INDEX')) {
```
```suggestion
                    if ('DROP' === $category && 0 !== strpos($command, 'DROP INDEX')) {
```

ausi replied:
Wouldn’t this miss all `ALTER TABLE … DROP INDEX …` queries?

aschempp replied:
True

aschempp commented on core-bundle/src/DependencyInjection/Compiler/TaggedMigrationsPass.php:
```diff
@@ -0,0 +1,51 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function process(ContainerBuilder $container): void
```
You should probably use the [PriorityTaggedServiceTrait](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/DependencyInjection/Compiler/PriorityTaggedServiceTrait.php)

ausi replied:
I don’t think that’s possible because we sort by priority first and by class name second.

aschempp commented on core-bundle/src/Migration/AbstractMigration.php:
```diff
@@ -0,0 +1,34 @@
+{
+    public function getName(): string
+    {
+        return str_replace('\\', ' ', static::class);
```
The default name is `Terminal42 FooBarBundle Migration SomeMigration`?

ausi replied:
Yes. Can you think of a better default name that identifies the migration unambiguously?

aschempp replied:
Maybe just the class name, then its obvious there‘s no name 🤷‍♂️

aschempp commented on core-bundle/src/Migration/AbstractMigration.php:
```diff
@@ -0,0 +1,34 @@
+        return str_replace('\\', ' ', static::class);
+    }
+
+    public function createResult(bool $successful = true, string $message = null): MigrationResult
```
shouldn't this method be `protected`? I also find it confusing that I can create a result without giving any arguments. Maybe `$successful` shouldn't have a value…

ausi replied:
Agree, will fix.

ausi replied:
Done in 47854333bba2587d8477bede3c8c3610e9a1b237

aschempp commented on core-bundle/src/Migration/MigrationResult.php:
```diff
@@ -0,0 +1,42 @@
+
+namespace Contao\CoreBundle\Migration;
+
+class MigrationResult
```
should this data object be `final`?

ausi replied:
Sounds reasonable to me to mark value objects `final`, @contao/developers WDYT?

ausi replied:
Done in cfb6404161f2a504c4e23af55cca52034307492a

aschempp commented on core-bundle/src/Command/MigrateCommand.php:
```diff
@@ -0,0 +1,278 @@
+    /**
+     * @var ?Installer
+     */
+    private $installer;
```
did you see there is an `Installer` in the core-bundle as well? Not exactly sure why, but I think we moved it from the installation bundle to be always available. Or maybe to opposite 🙈 🙃

ausi replied:
I don’t think they do exactly the same.

aschempp commented on core-bundle/src/Migration/MigrationInterface.php:
```diff
@@ -0,0 +1,22 @@
+
+interface MigrationInterface
+{
+    public function getName(): string;
```
would it be a problem if multiple migrations have the same name?

ausi replied:
Yes, as the names are listed when the command asks you if you want to execute the migrations.

aschempp replied:
well that's just unfortunate. But technically it won't be a problem? Like we're not storing the name of whether something was already executed?

ausi replied:
Correct. Technically duplicate names would not cause issues.

aschempp commented on installation-bundle/src/Database/Version330Update.php:
```diff
@@ -12,17 +12,35 @@
      * {@inheritdoc}
      */
-    public function shouldBeRun(): bool
+    public function shouldRun(): bool
```
can we keep a method alias and forward the call? I currently use these classes to execute Contao updates in doctrine migrations.

ausi replied:
Agree, will fix.
*(but your migrations probably will still need to be updated because of other changes to these classes)*

aschempp replied:
_I wasn't talking about **my** migrations, but about executing the Contao core migrations from Doctrine migrations. The constructor arguments are the same, so I don't see any changes 🙃_

ausi replied:
With *“Your migrations”* I was referring to your doctrine migrations 🙃
But you are right, I forgot that I kept the constructor the same.

aschempp reviewed (changes requested):
Looks very good and very simple! 🎉 

- I dislike the `Migrations` class name, it's very similar to `Migration` (the singular). Maybe it should be `MigrationsRunner` or `MigrationsHandler` or `…Manager`. I know these suck too, but I find the short name very confusing.
- Should we add a deprecation notice to the `AbstractVersionUpdate` class?
- as a lot of migrations depend on each other, are we sorting by name or how do we make sure they are executed in the correct order (except the service priority)?

ausi commented:
> I dislike the `Migrations` class name, it's very similar to `Migration` (the singular). Maybe it should be `MigrationsRunner` or `MigrationsHandler` or `…Manager`. I know these suck too, but I find the short name very confusing.

Of the ones listed I like `Migrations` best.

> Should we add a deprecation notice to the `AbstractVersionUpdate` class?

Agree, will fix.

> as a lot of migrations depend on each other, are we sorting by name or how do we make sure they are executed in the correct order (except the service priority)?

Yes, we sort by priority first and class name second, see https://github.com/contao/contao/pull/709/files#diff-d256c0387e834f3ac2078c93c9a88bc7R40-R47

Toflar commented:
I don't like `Migrations` either. Imho this is a `MigrationCollection` and the method should be named `runAll()`.

ausi committed:
Display how much migrations got executed

ausi committed:
Rename complete option to with-deletes

ausi committed:
Remove unnecessary default values and make createResult() protected

ausi committed:
Make MigrationResult final

ausi committed:
Deprecate AbstractVersionUpdate as it no longer gets used

ausi committed:
Rename and restructure Migrations to MigrationCollection

ausi committed:
Coding style

ausi commented:
> I don't like `Migrations` either. Imho this is a `MigrationCollection` and the method should be named `runAll()`.

Renamed in fc5219067a35ce9e67533160a40400ce9fd2d3b1. `runAll()` would be misleading IMO because only pending migrations get exectued, not all. I renamed to `run()` now, `runPending()` would be OK too I think.

> Should we add a deprecation notice to the `AbstractVersionUpdate` class?

Done in 23ec58b556a464bfcb4df05105fde987e04455d7.

I think all comments should be addressed now. @aschempp please take a look at your review comments and mark them as resolved if you agree with my changes.

ausi committed:
Fix left-out rename

ausi commented:
> please also provide a PR for the documentation :)

Docs pull request: https://github.com/contao/docs/pull/205

ausi committed:
Merge branch 'master' into feature/migrate-command

aschempp reviewed (dismissed):
I would suggest to use the FQCN instead of the generated name, but I‘m also fine with merging as is 👍🏻

ausi committed:
Use full class name as default name

ausi committed:
Merge branch 'master' into feature/migrate-command

ausi commented:
> I would suggest to use the FQCN instead of the generated name

Done in a6d2b26fe166a83d8e88808b7c88d705a36c0816

leofeyer commented on installation-bundle/src/Resources/config/services.yml:
```diff
@@ -2,12 +2,35 @@ services:
+
+    Contao\InstallationBundle\Database\:
+        resource: '../../Database/Version*'
+        exclude: '../../Database/Version{480,447}Update.php'
```
I guess we cannot use autowiring to make this a little less ugly?

ausi replied:
I don’t think so. `Version480Update` needs the `%kernel.project_dir%` which is just a string and AFAIK strings cannot be autowired?

bytehead replied:
They can be autowired:
```yml
# services.yml
services:
    _defaults:
        bind:
            $projectDir: '%kernel.project_dir%'
```

Toflar replied:
No autowiring in bundles ☝️ This here is "automatically loading classes" and not related to autowiring at all. So it's good as it is imho.

leofeyer replied:
I disagree. "No autowiring in bundles" in generally correct, however, this is a valid exception case IMHO, because loading version classes is a kind of closed sub-system within the application. And autowiring would be handy here.

But since this is legacy code that should be moved into the respective bundle anyway, I am fine with leaving it as is.

Toflar replied:
Well...maybe we should distinguish here. I'm not saying autowiring is always evil for bundles. We just have to keep in mind that the `core-bundle` is being extended by third party developers and autowiring would just make things possibly inconsistent depending on the application setup.
However, in this case, we could extract the migrations into their own `migrations.yaml` file in which case they would not affect the rest of the services:

```yaml
services:
    _defaults:
        autowire: true
        bind:
            string $projectDir: '%kernel.project_dir%'
    Contao\InstallationBundle\Database\:
        resource: '../../Database/Version*'
```

leofeyer replied:
Yes, that is what I had in mind. But as I said, it is fine for me the way it is as well.

leofeyer committed:
Fix the coding style

leofeyer committed:
Merge branch 'master' into feature/migrate-command

leofeyer commented:
Thank you @ausi.

Toflar commented:
This is biiiiig!! ❤️ 🎉

bytehead commented:
Thank you big time @ausi! 😍
